### PR TITLE
(Part of) API standardization

### PR DESCRIPTION
**Pull request by FBruzzesi**
# Description

Addresses the first and third bullet points of #590:

- Conform input features attributes to `n_features_in_` and adds warning for current attributes
- Moves class level attributes as class attributes and add warning for current

## Type of change
This is a Breaking change but retro-compatible with warning

# Checklist:

- [X] My code follows the style guidelines (flake8)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (also to the readme.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added tests to check whether the new feature adheres to the sklearn convention
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: sklego/linear_model.py [hunk 1]
```diff
@@ -59,9 +60,7 @@ def fit(self, X, y):
         return self
 
     def _calc_wts(self, x_i):
-        distances = np.array(
-            [np.linalg.norm(self.X_[i, :] - x_i) for i in range(self.X_.shape[0])]
-        )
+        distances = np.linalg.norm(self.X_ - x_i, axis=1)
```
Comment (FBruzzesi) [inline sklego/linear_model.py:63]: Some very tiny optimization avoiding for loops

File: sklego/linear_model.py [hunk 2]
```diff
@@ -76,9 +75,10 @@ def predict(self, X):
         """
         X = check_array(X, estimator=self, dtype=FLOAT_DTYPES)
         check_is_fitted(self, ["X_", "y_"])
-        results = np.zeros(X.shape[0])
-        for idx in range(X.shape[0]):
-            results[idx] = np.average(self.y_, weights=self._calc_wts(x_i=X[idx, :]))
+
+        results = np.stack(
+            [np.average(self.y_, weights=self._calc_wts(x_i=x_i)) for x_i in X]
+        )
```
Comment (FBruzzesi) [inline sklego/linear_model.py:81]: Another very tiny optimization avoiding for loops

File: sklego/model_selection.py [hunk 3]
```diff
@@ -643,14 +643,3 @@ def _regroup(self, groups):
...
-    def _method_is_fitted(self, X):
-        """
-        :param X: Array to use if the method is fitted
-        :return: True if fitted, else False
-        """
-        try:
-            self.cluster_method.predict(X[0:1, :])
-            return True
-        except NotFittedError:
-            return False
```
Comment (FBruzzesi) [inline sklego/model_selection.py:656]: Looks like this method was copy-pasted from `KlusterFoldValidation`, and requires `cluster_method` which is not available for `GroupTimeSeriesSplit`

File: sklego/preprocessing/patsytransformer.py [hunk 4]
```diff
@@ -24,7 +24,12 @@ def fit(self, X, y=None):
 
         # check the number of observations hasn't changed. This ought not to
         # be necessary given NA_action='raise' above but just to be safe
-        assert np.array(X_).shape[0] == np.array(X).shape[0]
```
Comment (FBruzzesi) [inline sklego/preprocessing/patsytransformer.py:27]: `assert` should be avoided in production code as running non-debug mode would skip that anyway

**Review by koaning — APPROVED** [review]
Whoa. That's a bunch of little things cleaned up. Nicely done :)

**Comment by koaning** [comment]
I think this PR is good to merge as-is.

**Pull request merged**